core/common/privdata/membershipinfo.go

```diff
@@ -7,36 +7,97 @@ SPDX-License-Identifier: Apache-2.0
+	mspp "github.com/hyperledger/fabric/protos/msp"
+	"github.com/pkg/errors"
 )
 
 var logger = flogging.MustGetLogger("common.privdata")
 
 // MembershipProvider can be used to check whether a peer is eligible to a collection or not
 type MembershipProvider struct {
+	mspID                       string
 	selfSignedData              common.SignedData
 	IdentityDeserializerFactory func(chainID string) msp.IdentityDeserializer
 }
 
 // NewMembershipInfoProvider returns MembershipProvider
-func NewMembershipInfoProvider(selfSignedData common.SignedData, identityDeserializerFunc func(chainID string) msp.IdentityDeserializer) *MembershipProvider {
-	return &MembershipProvider{selfSignedData: selfSignedData, IdentityDeserializerFactory: identityDeserializerFunc}
+func NewMembershipInfoProvider(mspID string, selfSignedData common.SignedData,
+	identityDeserializerFunc func(chainID string) msp.IdentityDeserializer) *MembershipProvider {
+	return &MembershipProvider{mspID: mspID, selfSignedData: selfSignedData, IdentityDeserializerFactory: identityDeserializerFunc}
 }
 
 // AmMemberOf checks whether the current peer is a member of the given collection config.
 // If getPolicy returns an error, it will drop the error and return false - same as a RejectAll policy.
-func (m *MembershipProvider) AmMemberOf(channelName string, collectionPolicyConfig *common.CollectionPolicyConfig) (bool, error) {
+// It is used when a chaincode is upgraded to see if the peer's org has become eligible after	a collection
+// change.
+func (m *MembershipProvider) AmMemberOf(channelName string,
+	collectionPolicyConfig *common.CollectionPolicyConfig) (bool, error) {
 	deserializer := m.IdentityDeserializerFactory(channelName)
+
+	// Do a simple check to see if the mspid matches any principal identities in the SignaturePolicy - FAB-17059
+	if collectionPolicyConfig.GetSignaturePolicy() != nil {
+		memberOrgs, err := getMemberOrgs(collectionPolicyConfig.GetSignaturePolicy().GetIdentities(), deserializer)
+		if err != nil {
+			logger.Errorf("Failed getting member orgs from MSP Principals: %s", err)
+		}
+
+		if _, ok := memberOrgs[m.mspID]; ok {
+			return true, nil
+		}
+	}
+
+	// Fall back to default access policy evaluation otherwise
 	accessPolicy, err := getPolicy(collectionPolicyConfig, deserializer)
 	if err != nil {
 		// drop the error and return false - same as reject all policy
 		logger.Errorf("Reject all due to error getting policy: %s", err)
 		return false, nil
 	}
+
 	if err := accessPolicy.Evaluate([]*common.SignedData{&m.selfSignedData}); err != nil {
 		return false, nil
 	}
+
 	return true, nil
 }
+
+// getMemberOrgs returns a map containing member orgs from a list of MSPPrincipals
+func getMemberOrgs(identities []*mspp.MSPPrincipal, deserializer msp.IdentityDeserializer) (map[string]struct{}, error) {
+	memberOrgs := make(map[string]struct{})
+
+	// get member org MSP IDs from the envelope
+	for _, principal := range identities {
```

Comment:
Yep, sure I can move it to a shared place. I left it separate because we ended up changing it to return a map instead of a string slice. I can change the consumers of SimpleCollection (coordinator) to also check the map instead.